extensions/amp-access-scroll/0.1/read-depth-tracker.js

```diff
@@ -0,0 +1,110 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Services} from '../../../src/services';
+import {debounce} from '../../../src/utils/rate-limit';
+
+export class ReadDepthTracker {
+  /**
+   * Sets up tracking of paragraphs in document.
+   * @param {!../../../src/service/ampdoc-impl.AmpDoc} ampdoc
+   * @param {!../../amp-access/0.1/amp-access-source.AccessSource} accessSource
+   * @param {string} connectHostname
+   */
+  constructor(ampdoc, accessSource, connectHostname) {
+    /** @private {!../../../src/service/ampdoc-impl.AmpDoc} */
+    this.ampdoc_ = ampdoc;
+
+    /** @private {number?} */
+    this.lastReadIndex_ = null;
+
+    /** @private {!../../amp-access/0.1/amp-access-source.AccessSource} */
+    this.accessSource_ = accessSource;
+
+    /** @private {string} */
+    this.connectHostname_ = connectHostname;
+
+    /** @private {function()} */
+    const debouncedFindTopParagraph_ = debounce(
+        ampdoc.win,
+        this.findTopParagraph_.bind(this),
+        1000
+    );
+
+    this.viewport_ = Services.viewportForDoc(ampdoc);
+    this.viewport_.onChanged(debouncedFindTopParagraph_);
+
+    this.paragraphs_ = ampdoc.getBody().getElementsByTagName('p');
```

Comment:
One nuances here is that we try to initial and authorize as early as possible. Even if the document is not fully parsed yet. That means that this query may return an incomplete list of `<p>` tags.

Due to this, I recommend making this a promise chained on `ampdoc.whenReady()` promise - that will ensure that the document is fully parsed before you search for paragraphs.

Reply:
Nevermind. I missed the part that DOM is fully parsed by this point. Canceling this request.

Reply:
Bringing this back. I was right the first time. PTAL. We need to wait for DOM to be parsed if we want all `<p>` tags.

Reply:
@dvoytenko thanks for looking into this. I think it also makes sense to wait for the document to start checking for the top paragraph. Let me know if this change looks okay.

Reply:
@helin24 Thank you for the change, this looks correct.